src/MongoDB.Driver/PipelineDefinition.cs

```diff
@@ -332,29 +332,26 @@ public override RenderedPipelineDefinition<TOutput> Render(IBsonSerializer<TInpu
 
         private static List<IPipelineStageDefinition> VerifyStages(List<IPipelineStageDefinition> stages)
         {
-            var nextInputType = typeof(TInput);
+            var expectedInputType = typeof(TInput);
             for (int i = 0; i < stages.Count; i++)
             {
-                if (stages[i].InputType != nextInputType)
+                if (stages[i].InputType != expectedInputType)
                 {
-                    var message = string.Format(
-                        "The input type to stage[{0}] was expected to be {1}, but was {2}.",
-                        i,
-                        nextInputType,
-                        stages[i].InputType);
-                    throw new ArgumentException(message, "stages");
+                    var message =
+                        $"The input type to stage[{i}] was expected to be {expectedInputType}, but was {stages[i].InputType}.";
+                    throw new ArgumentException(message, nameof(stages));
                 }
 
-                nextInputType = stages[i].OutputType;
+                expectedInputType = stages[i].OutputType;
             }
 
-            if (nextInputType != typeof(TOutput))
+            var outputType = stages.Last().OutputType;
```

Comment:
What you have is fine and you don't have to change it if you like what you have more, but I would have used:

    private static List<IPipelineStageDefinition> VerifyStages(List<IPipelineStageDefinition> stages)
    {
        var expectedInputType = typeof(TInput);
        for (int i = 0; i < stages.Count; i++)
        {
            if (stages[i].InputType != expectedInputType)
            {
                var message =
                    $"The input type to stage[{i}] was expected to be {expectedInputType}, but was {stages[i].InputType}.";
                throw new ArgumentException(message, nameof(stages));
            }

            expectedInputType = stages[i].OutputType;
        }
        var lastStageOutputType = expectedInputType; 

        if (lastStageOutputType != typeof(TOutput))
        {
            var message =
                $"The output type to the last stage was expected to be {typeof(TOutput)}, but was {lastStageOutputType}.";
            throw new ArgumentException(message, nameof(stages));
        }

        return stages;
    }

The following is more efficient than using LINQ:

    var lastStageOutputType = expectedInputType; 

because it's a simple assignment statement.